timemachine/md/exchange/exchange_mover.py

```diff
@@ -51,6 +53,28 @@ def translate_coordinates(coords, new_loc):
     return centered_coords + new_loc
 
 
+@dataclass(frozen=True)
+class WaterSamplingDiagnostics:
+    proposals_by_state_by_iter: NDArray
+    replica_idx_by_state_by_iter: list[list[ReplicaIdx]] | None = None
```

Comment:
It seems redundant to include this here, since it's already in `HREXDiagnostics`. I think it'd make more sense to me for `proposals_by_replica_by_iter` to be a method accepting a `replica_idx_by_state_by_iter` argument, or to rely on a general utility to invert permutations, since we now have a few instances of functions from `x_by_state -> x_by_replica`.

Reply:
> Could you write a general utility for the permutations? Would be very handy.

I had something like this started in https://github.com/proteneer/timemachine/pull/1508, but ran into some minor annoyances making it compatible with both JAX and numpy arrays. Will plan to revisit.